### PR TITLE
afew: add the 'filters' option

### PR DESCRIPTION
**Pull request by aheaume**
Hello!

This PR adds a new `filters` option to the afew module. This new option allows one to write filter declarations in a more structured way. See the `filters.example` and `extraConfig.example` configuration options.

It is backward compatible, no changes to existing configurations are necessary.

Would you be interested in merging this into master?

Have a nice day,
Alex

### CODE REVIEW

**Comment by bsima** [comment]
Nice! I've applied your patch locally and it seems to be working well.

File: modules/programs/afew.nix [hunk 1]
```diff
@@ -6,21 +6,86 @@ let
...
+    in
+      concatStringsSep "\n" (map mkSection afewFilters);
+
 in
 
 {
   options.programs.afew = {
     enable = mkEnableOption "the afew initial tagging script for Notmuch";
 
+    filters = mkOption {
```
Comment (rycee) [inline modules/programs/afew.nix]: I find this a bit tricky to follow. Would it be possible to express it with a number of separate options instead? Like `programs.afew.filters.spam`, `programs.afew.filters.killThreads`, etc.

So the example given for this option would instead be
```
programs.afew = {
  filters = {
    spam = [
      {
        query = "frompointyheaded@boss.com";
        tags = ["-new" "+boss"];
        message = "Message from above";
      }
    ];

    inbox = {};
  };
}
```

Comment (aheaume) [reply modules/programs/afew.nix]: Thanks for the input :)

In your example, the "frompointyheaded@boss.com" filter is a generic filter (not part of the spam filter), and there maybe any number of generic filters like this one. So it would be more like:

```
programs.afew = {
  filters = {
    spam = {};

    filter = {
      query = "frompointyheaded@boss.com";
      tags = ["-new" "+boss"];
      message = "Message from above";
    };

    filter = {
      query = "somebodyelse@boss.com";
      tags = ["-new" "+boss"];
      message = "Another message from above";
    };

    inbox = {};
  };
}
```

Unfortunately, filters in a few are applied in order which introduces dependencies. As I understand, attribute sets in nix are ordered alphabetically, so `filters` needs to be a list. I'm very new to nix so I could be wrong on this one :) I'm happy to change this PR if you know of a way to bend the way nix orders attributes in an attribute set.

**Comment by sboosali** [comment]
iiuc, this works (with `lib.types.listOf`, and `lib.types.submodule` for
query/tage/message):


```

programs.afew = {
 spamFilter = {};

 inboxFilter = {};

 filters = [
   {
     query = "frompointyheaded@boss.com";
     tags = ["-new" "+boss"];
     message = "Message from above";
   }

   {
     query = "somebodyelse@boss.com";
     tags = ["-new" "+boss"];
     message = "Another message from above";
   }

 ]


 };
}

```

**Comment by aheaume** [comment]
Interesting. I initially tried something along these lines, but `spamFilter` and `inboxFilter` (and all the built-in filters afew provides) are like regular filters that need to be part of the `filters` list as well. Furthermore, afew can be extended with new filter classes (hence the `class` attribute in my patch, which defaults to the generic `Filter` when absent).

**Comment by aheaume** [comment]
Maybe it is my original example that is a bit misleading. Here is another, more complete one.

```
# Built-in Filter, needs to be parameterized
{
  class = "Filter" # can be omitted since `class` defaults to Filter
  query = "fromspammyboss@boss.com";
  tags = ["+boss" "+spam"];
  # no message
}

# Built-in SpamFilter, does not need to be parameterized, runs before other filters
{ class = "SpamFilter"; }

# Built-in Filter, needs to be parameterized
{
  class = "Filter" # can be omitted since `class` defaults to Filter
  query = "frompointyheaded@boss.com";
  tags = ["-new" "+boss"];
  message = "Message from above"; }

# Built-in InboxFilter
{ class = "InboxFilter"; }

# Custom filter defined in ~/.config/afew/AfterInbox.py, depends on the "frompointyheaded" and "fromspammyboss" Filters and InboxFilter (for the sake of the example)
# Custom filters are used to run arbitrary code for when Filter isn't enough
{
  class = "AfterInbox"
  query = "tag:boss"
  another_param = "bar"
  tags = ["-boss" "-inbox" "+spam"]
}
```

I hope that this examples clarifies why order is important, and why built-in filters like SpamFilter and InboxFilter can't be treated apart from other filters.

The afew configuration file is really a programming-language-in-ini-format which can't easily be represented with clean attribute sets, unfortunately. I'd love to be mistaken and improve this PR though :)

Part of my motivation for this PR is to be able to use the nix expression language to configure afew, instead of the ini format.

**Comment by rycee** [comment]
Ya, I completely misunderstood the way these filters work! With your clarification I think the design makes sense.

My one concern would be configuration merging. With the current design merging would not necessarily make sense. Imagine you want to have an afew-configuration where you have `afew-common.nix`, `afew-work.nix`, and `afew-private.nix` where the work and private configurations `import` the common configuration. In the common configuration you may have a "spam" and "inbox" filter while in the work filter you have the pointy hair stuff that should go in between the "spam" and "inbox" filters. The list merging wouldn't allow such a setup.

So I wonder if not the ideal would be to use a DAG here. Then you could have
```
{
  spam = dag.entryAnywhere { class = "SpamFilter"; };
  inbox = dag.entryAfter ["spam"] { class = "InboxFilter"; };
}
```
in `afew-common.nix` and in `afew-work.nix`
```
{
  bossFilter = dag.entryBetween ["spam"] ["inbox"] {
    query = "frompointyheaded@boss.com";
    tags = ["-new" "+boss"];
    message = "Message from above"; 
  };
}
```

That said, this complicates the setup further so I think using a plain list will be OK for now.

File: modules/programs/afew.nix [hunk 2]
```diff
@@ -43,10 +108,10 @@ in
     home.packages = [ pkgs.afew ];
 
     xdg.configFile."afew/config".text = ''
-      # Generated by Home Manager.
-      # See https://afew.readthedocs.io/
-
+      # Generated by Home Manager.\n# See https://afew.readthedocs.io/\n\n
```
Comment (rycee) [inline modules/programs/afew.nix]: Why was this change needed?

Comment (aheaume) [reply modules/programs/afew.nix]: Not needed, I'll revert this.

**Comment by aheaume** [comment]
A DAG would indeed be very cool! I wonder if it would not be overkill though. Maybe we could add `preFilters` and `postFilters` right now in order to alleviate the order requirements? From anecdotal evidence, I gather that the ordering afew imposes is mostly used to speed things up by running the SpamFilter first to weed out mail that does not need to be tagged, and the InboxFilter last to ignore already-tagged mail during the next tagging process.

One thing that I've been thinking about is that my design effectively restrict users to ever use the `class` parameter name on their custom filters (as it would be processed and removed from the attribute set). Probably not that big of a deal, especially since on the python side of the custom filter the `class` keyword is probably impractical to use, but maybe this is an opportunity to remove this restriction and lay the groundwork for a DAG implementation in the future by providing more out-of-band data about the filters.

Combining the two, this would mean amending this PR now to support the following configuration:

```
preFilters = {
  spam = { class = SpamFilter" };
};

filters = {
  bossFilter = { 
    class = "Filter"; 
    parameters = {
      query = "frompointyheaded@boss.com";
      tags = ["-new" "+boss"];
      message = "Message from above";
    };
  };
};

postFilters = {
  inbox = { class = "InboxFilter" };
};
```

which removes the need for explicit ordering of the `filters`, and moves the `class` attribute out of the parameters set. This would definitely be enough for my own configuration.

Then, if the need for a DAG implementation arises, we could support:

```
preFilters = {
  spam = { class = SpamFilter" };
};

filters = {

  someFilter = {
    class = "Filter";
    parameters = { foo = "bar"; };
  };

  bossFilter = { 
    class = "Filter"; 
    parameters = {
      query = "frompointyheaded@boss.com";
      tags = ["-new" "+boss"];
      message = "Message from above"; 
    };
    after = ["someFilter"];
  };

};

postFilters = {
  inbox = { class = "InboxFilter" };
};
```

WDYT?

**Comment by aheaume** [comment]
I have been thinking about various possible improvements to the current code over the past week, and I believe the current list implementation is the best one as it is closer to the afew ini config file.

For the record, here are some implementations I considered

- the DAG idea is a good one, but would require a lot of explicit ordering on the part of the user
- the pre/filters/post is good enough, but does not support complex dependencies
- a new "stages" approach, a generalization of the pre/filters/post, where stages would be explicitly ordered (`["pre" "filters" "post"]` by default) and filters could be assigned to one of the stages (`filters` by default), but this introduces a lot of overhead on the part of the user, and tagging mail is probably not worth the trouble.

@rycee if you are OK, I believe we could work towards merging and closing this PR.

**Comment by rycee** [comment]
Fair enough! I've added a commit with a few minor cleanups. If they are working OK for you then I'll merge as-is. I tried the example setup and get
```ini
[SpamFilter.1000]

[Filter.1001]
message=Message from above
query=frompointyheaded@boss.com
tags=-new;+boss

[InboxFilter.1002]
```
which I guess is as expected?

**Comment by aheaume** [comment]
This is the expected output. I confirm that it is working for me with your changes. LGTM!

**Comment by aheaume** [comment]
@rycee Actually, I may have found a way to make the DAG work that would not introduce too much configuration overhead.

The main issue I had with the DAG was that a handful of filters need to run before anything else, and another handful need to run after everything else. By introducing a skeleton of dummy nodes in the graph, this is in fact easy.

```
# skeleton
begin = dag.entryAnywhere { ... }
filters = dag.entryAfter [ "begin" ] { ... }
last = dag.entryAfter [ "filters" ] { ... }

# begin
spam = dag.entryAfter [ "begin" ] { class = "SpamFilter"; }
spam = dag.entryAfter [ "begin" ] { class = "KillThreadsFilter"; }

# filters
someFilter = dag.entryAfter [ "filters" ] { class = "Filter"; parameters = { foo = "bar"; }; }
otherFilter = dag.entryAfter [ "someFilter" ] { class = "Filter"; parameters = { foo = "bar"; }; }

# last
inbox = dag.entryAfter [ "last" ] { class = "InboxFilter"; }
```

I need to figure out what the implementation of the dummy node is, I guess a dummy filter that just outputs a message to the afew log would be OK.
It seems to me that this would work in the configuration merging scenario (by just grafting after the appropriate skeleton nodes.)
If that looks OK, then maybe don't merge this PR yet :) I will likely have the time to work on this in the coming weeks.

File: modules/programs/afew.nix [hunk 3]
```diff
@@ -6,21 +6,88 @@ let
...
+      description = ''
+        Filters added to afew configuration file. Available configuration
+        options are described in the afew manual: <link
+        xlink:href="https://afew.readthedocs.io/en/latest/configuration.html" />.
+      '';
+      example = literalExample ''
+        [
+          { class = "SpamFilter"; }
+
+          { query = "frompointyheaded@boss.com";
```
Comment (vdemeester) [inline modules/programs/afew.nix]: There is a typo here : `from:pointyheaded@boss.com`

**Comment by rycee** [comment]
To further simplify the configuration (but complicate the implementation) perhaps it would be possible to replace
```
someFilter = dag.entryBetween [ "filters" ] [ "last" ] {
  class = "Filter";
  parameters = { foo = "bar"; };
}
```
with something like
```
someFilter = {
  after = [ "filters" ];
  before = [ "last" ];
  class = "Filter";
  parameters = { foo = "bar"; };
}
```
and then apply the dag functions internally? I.e.
```nix
let
  dagify = filter:
    dag.entryBetween filter.before filter.after (removeAttrs filter ["after" "before"]);
in
  map toFilterIni (dag.topoSort (mapAttrs dagify cfg.filters))
```

**Comment by aheaume** [comment]
I have pushed a DAG-based implementation of the `filters` option. It is ready for review if anybody has some extra time on their hands :) I've had excellent results on some test data.

There are a few things still missing though:

- the `extraConfig` option must be kept around as I don't want to break existing configurations, but it clashes with the new `filters` option, what is the recommended way to disable `extraConfig` when `filters` is used?

- Error handling when the DAG has cycles/loops, "missing attribute result" is likely to be a bit puzzling

- Porting my real configuration over to the new format and check everything is OK.

File: modules/programs/afew.nix [hunk 4]
```diff
@@ -6,21 +6,134 @@ let
...
+    in
+      concatStringsSep "\n" (imap0 mkSection filters);
+
 in
 
 {
   options.programs.afew = {
     enable = mkEnableOption "the afew initial tagging script for Notmuch";
 
+    defaultBefore = mkOption {
```
Comment (rycee) [inline modules/programs/afew.nix]: Are the `defaultBefore`, `defaultAfter`, and `defaultFilters` options necessary? I figure these could simply be hard coded?

File: modules/programs/afew.nix [hunk 5]
```diff
@@ -6,21 +6,134 @@ let
...
+      };
+      description = ''
+        An initial value for the filters DAG. It should only contain
+        "dummy" nodes. Those dummy nodes are conventionally prefixed
+        with @.
+      '';
+    };
+
+    filters = mkOption {
+      type = types.attrsOf types.attrs;
```
Comment (rycee) [inline modules/programs/afew.nix]: I think the type here could be changed as per
```suggestion
      type = with types;
        let
          primitive = either bool (either int str);
        in
          attrsOf (attrsOf (either primitive (listOf str)));
```
Perhaps `primitive` could even be `primitive = either bool str`?

File: modules/programs/afew.nix [hunk 6]
```diff
@@ -6,21 +6,134 @@ let
...
+
+  /* Return a topologically sorted list of DAG entries encapsulating
+     name value pairs.
+  */
+  reorderFilters = filters:
+    (dag.topoSort (mkDag filters)).result;
+
+  /* Convert a topologically sorted list of DAG nodes to a list of INI
+     sections.  The section name is appended with an index number to
+     prevent clashes, as is required by afew.
```
Comment (rycee) [inline modules/programs/afew.nix]: Is the index number still necessary? Seems to me that the attribute name is used now and that should by definition be unique.

File: modules/programs/afew.nix [hunk 7]
```diff
@@ -6,21 +6,134 @@ let
...
+            class = "InboxFilter";
+            before = [];
+            after = ["@end"];
+          };
+        }
+      '';
+    };
+
     extraConfig = mkOption {
       type = types.lines;
```
Comment (rycee) [inline modules/programs/afew.nix]: I'd suggest to change this to `types.nullOr types.lines` and the default to `null`. Then add
```
warnings = optional (cfg.extraConfig != null)
  "'programs.afew.extraConfig' is deprecated, use 'programs.afew.filters' instead.";
```
and change
```
      ${cfg.extraConfig}
      ${toOrderedINI { mkKeyValue = mkIniKeyValue; } (reorderFilters (cfg.defaultFilters // cfg.filters))}
```
to
```
      ${
        if (cfg.extraConfig != null)
        then cfg.extraConfig
        else
          toOrderedINI { mkKeyValue = mkIniKeyValue; }
          (reorderFilters
          (defaultFilters // cfg.filters))
      }
```
where `defaultFilters` should be `cfg.defaultFilters` if it is not possible to hard code them.

Perhaps this will allow sufficient backwards compatibility and won't cause issues of odd mixing of configurations.

File: modules/programs/afew.nix [hunk 8]
```diff
@@ -6,21 +6,134 @@ let
...
+    defaultFilters = mkOption {
+      type = types.attrsOf types.attrs;
+      default = {
+        "@begin" = { message = "BEGIN"; after = []; before = []; };
+        "@end" = { message = "END"; after = ["@begin"]; before = []; };
+      };
+      description = ''
+        An initial value for the filters DAG. It should only contain
+        "dummy" nodes. Those dummy nodes are conventionally prefixed
+        with @.
```
Comment (rycee) [inline modules/programs/afew.nix]:
```suggestion
        with <literal>@</literal>.
```

File: modules/programs/afew.nix [hunk 9]
```diff
@@ -6,21 +6,134 @@ let
...
+        in
+          dag.entryBetween before after (nameValuePair filterName filterKV);
+    in
+      mapAttrs mkNode filters;
+
+  /* Return a topologically sorted list of DAG entries encapsulating
+     name value pairs.
+  */
+  reorderFilters = filters:
+    (dag.topoSort (mkDag filters)).result;
```
Comment (rycee) [inline modules/programs/afew.nix]: If you leave this as
```
  reorderFilters = filters: dag.topoSort (mkDag filters);
```
then you could have an assertion
```
assertions = [
  {
    assertion = reorderFilters ? result;
    message = "The 'programs.afew.filters' option contains a circular dependency.";
  }
];
```

**Comment by aheaume** [comment]
> Sorry about the delay!

No worries :) I've been meaning to work on this but can't find the time at the moment.

File: modules/programs/afew.nix (hunk 6 above)
Comment (aheaume) [reply modules/programs/afew.nix]: The index number is used to prevent clashes. See [here](https://afew.readthedocs.io/en/latest/filters.html#headermatchingfilter) for an example in afew documentation. Unfortunately I don't think we can get rid of it as long as a few requires it.

File: modules/programs/afew.nix (hunk 4 above)
Comment (aheaume) [reply modules/programs/afew.nix]: I will hardcode them for now to simplify the current implementation, and we'll see if anybody complains about the lack of flexibility.

File: modules/programs/afew.nix (hunk 5 above)
Comment (aheaume) [reply modules/programs/afew.nix]: As afew can be user-extendable, I'm not sure what the allowed types of the leaves should be.
I think we could go with `(either primitive (listOf primitive))`.

File: modules/programs/afew.nix (hunk 7 above)
Comment (aheaume) [reply modules/programs/afew.nix]: This looks great, however I'm not familiar enough with the nix expression language or home-manager to know how to use `warnings`. Can you point me to the documentation or an example?

File: modules/programs/afew.nix (hunk 9 above)
Comment (aheaume) [reply modules/programs/afew.nix]: Same as above with `warnings`, I'm not sure how to use `assertions`.
